test/migration/migration_test.go

```diff
@@ -0,0 +1,219 @@
+
+func TestMain(m *testing.M) {
+	kubeconfig = flag.String("kubeconfig", "", "kube config path, e.g. $HOME/.kube/config")
+	opImage = flag.String("operator-image", "", "operator image, e.g. quay.io/coreos/prometheus-operator")
+	ns = flag.String("namespace", "prometheus-operator-e2e-tests", "e2e test namespace")
+	flag.Parse()
+
+	os.Exit(m.Run())
+}
+
+func TestMigration(t *testing.T) {
+	// Create all the clients required.
+	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	kclient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	mv, err := k8sutil.GetMinorVersion(kclient.Discovery())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if mv < 7 {
+		t.Skip("lower than 1.7 version")
+		return
+	}
+
+	extClient, err := apiextensionsclient.NewForConfig(config)
+	if err != nil {
+		t.Fatal(err)
+	}
+	monClient, err := v1alpha1.NewForConfig(config)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tprClient := kclient.ExtensionsV1beta1().ThirdPartyResources()
+	crdClient := extClient.ApiextensionsV1beta1().CustomResourceDefinitions()
+
+	if framework, err = operatorFramework.New(
+		*ns,
+		*kubeconfig,
+		"quay.io/coreos/prometheus-operator:v0.10.2",
+	); err != nil {
+		log.Printf("failed to setup framework: %v\n", err)
+		t.Fatal(err)
+	}
+
+	ctx := framework.NewTestCtx(t)
+	defer ctx.Cleanup(t)
+	ns2 := ctx.CreateNamespace(t, framework.KubeClient)
+	ctx.SetupPrometheusRBAC(t, ns2, framework.KubeClient)
+
+	// Launch the objects.
+	name := "test"
+	group := "servicediscovery-test"
+
+	prometheusTPR := framework.MakeBasicPrometheus(ns2, name, name, 1)
+	prometheusTPR.Namespace = ns2
+	if err := framework.CreatePrometheusAndWaitUntilReady(ns2, prometheusTPR); err != nil {
+		t.Fatal(err)
+	}
+
+	s := framework.MakeBasicServiceMonitor(group)
+	if _, err := framework.MonClient.ServiceMonitors(ns2).Create(s); err != nil {
+		t.Fatal("Creating ServiceMonitor failed: ", err)
+	}
+
+	if err := framework.CreateAlertmanagerAndWaitUntilReady(ns2, framework.MakeBasicAlertmanager(name, 3)); err != nil {
+		t.Fatal(err)
+	}
+
+	obj, err := monClient.Prometheuses(ns2).List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	proms := obj.(*v1alpha1.PrometheusList)
+
+	// Get the objects.
+	obj, err = monClient.Alertmanagers(ns2).List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	ams := obj.(*v1alpha1.AlertmanagerList)
+
+	obj, err = monClient.ServiceMonitors(ns2).List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	sms := obj.(*v1alpha1.ServiceMonitorList)
+
+	// Delete and launch new operator.
+	deletePolicy := metav1.DeletePropagationForeground
+	if err := framework.
+		KubeClient.
+		Extensions().
+		Deployments(framework.Namespace.Name).
+		Delete("prometheus-operator", &metav1.DeleteOptions{
+			PropagationPolicy: &deletePolicy,
+		}); err != nil {
+		t.Fatal(err)
+	}
+
+	wait.Poll(time.Second, 3*time.Minute, func() (bool, error) {
```

Comment:
If `3*time.Minute` is not specifically tuned, can you use `framework.DefaultTimeOut`? This helps us in the future to be able to adjust the timeouts in one place for slow environments.